### ISSUE TITLE
Use in npm project

### ISSUE THREAD

**Issue by benbro** [jupyterlite/jupyterlite#947]
How can I use jupyterlite as part of a larger js project built with npm?
There are several [npm packages](https://www.npmjs.com/search?q=%40jupyterlite) but I couldn't find explanation in the docs which package I should include in my project and how to import and start the notebook.

**Comment by bollwyvl** [comment]
Right: 99% of the code that comprises the application is published under the `@jupyterlite/*` packages, and these can be confected in any way a JS toolchain can effectively service. A look at the [individual](https://github.com/jupyterlite/jupyterlite/blob/main/app/repl/index.template.js) [app](https://github.com/jupyterlite/jupyterlite/blob/main/app/retro/index.template.js) [templates](https://github.com/jupyterlite/jupyterlite/blob/main/app/lab/index.template.js) shows how one can build a new application from the core packages.

However, this has been left mostly undocumented, as the volunteer maintainers of this project are not going to be able service the _many_ inevitable issues that arise from the vast number of JS tool chains. At this point in the project's evolution, the effort is almost entirely budgeted on efficiently and cleanly implementing as many features of the upstream JupyterLab application as possible, and maintaining compatibility with labextensions that are built for the same, and (generally) distributed via the python ecosystem.

**Comment by benbro** [comment]
Thank for the link to the individual app templates. I'm trying to follow it.

How does [new JupyterLab({})](https://github.com/jupyterlite/jupyterlite/blob/main/app/lab/index.template.js#L197) knows where to render the notebook in the html? I don't see it in the code.

Where is [style.js](https://github.com/jupyterlite/jupyterlite/blob/main/app/lab/index.template.js#L11)?

**Comment by bollwyvl** [comment]
> where to render the notebook in the html

This is all managed by the `app.shell`. Of note: there's basically 0 chance of using the components as they exist today outside of the constraints of a full-page `@lumino/application:Application`: if anything, start with the `repl` example in an iframe, and try to build out from there.

> where is style.js

This is generated during the webpack build process, and requires a thorough look at the webpack configuration of both [lite](https://github.com/jupyterlite/jupyterlite/blob/main/app/webpack.config.js) and the [upstream](https://github.com/jupyterlab/jupyterlab/blob/master/builder/src/webpack.config.base.ts).

**Comment by benbro** [comment]
Using in a JS toolchain can mean two things:

1. Packaging all static assets needed so we could use npm instead of the python CLI.
2. Being able to import jupyterlite in a JS project, configure it and render it to custom element on the page.

Is it currently possible to do either? Are there plans or interest to support it?

Will it be possible to render jupyterlite inside an iframe and communicate with it from the main page using Window.postMessage()?

**Comment by bollwyvl** [comment]
> can mean two things

really more _an infinite number of things_: see other issues about react, vue, svelte, web components. i'm surprised the different versions of angular people haven't shown up yet and argued about their stuff.

>  possible to do either

no

>  render jupyterlite inside an iframe

yes, see https://github.com/jupyterlite/jupyterlite/issues/922

if you implement [`PageConfig`](https://jupyterlab.readthedocs.io/en/stable/api/modules/coreutils.pageconfig.html) (hint: you have to implement `PageConfig`) you can also configure  `jupyter-lite.json#/jupyter-config-data/exposeAppInBrowser` (see [docs](https://github.com/jupyterlab/jupyterlab/blob/v3.5.2/jupyterlab/labapp.py#L456), and get access to the top-level object with `iframe.contentWindow.jupyterapp`, but this has _additional_ security considerations.... sort of. Anybody with a browser console can get in there anyway, by looking at event handlers.

**Comment by benbro** [comment]
I understand that it's not possible to support all JS libraries but a demo in one of them will let the community expand it to all others. A  webpack project that imports jupyterlite, renders it and communicate with it can serve as a template to all other use cases. I'll be happy to work on it if you can help me get started.

Having an npm package with the static assets will be extremely useful. With this we could render to an iframe in any JS library (react, vue...) without having to manually copy the output of the python CLI.

**Comment by bollwyvl** [comment]
Again, see the comments and issues above: the linked templates _are_ the documentation at this point, as the API is still changing while in beta.

To expound on this: there is no "The JupyterLite," as it's three apps built entirely out of extensions, some from jupyterlab, some from this repo. It inherits the limitations of the upstream project in order to be able to share in benefits of integration with the rest of the jupyter ecosystem. The current configuration is optimized for _users_: this is informed by a five-year-long experiment of forcing every scientist and student who wanted to use jupyter tools to become nodejs/npm/webpack expert. Prebuilt, independently-packaged, single-package-version-per-name have drastically improved the situation.

A look at the demo configuration deployed on RTD shows a [fairly heavyweight smattering](https://github.com/jupyterlite/jupyterlite/blob/main/examples/jupyter_lite_config.json) of other extensions to actually deliver interesting content. _None_ of those extensions are installable in a usable form from `npm`, today... this would have to be taken up with the [upstream](https://github.com/jupyterlab/jupyterlab).

Back on this repo: anyone is free to PR a schema-compatible way of providing all of the features of the existing `jupyter lite` CLI on an alternate language runtime. Heck, one could even run `jupyterlite` inside `pyodide` inside `wasm` inside `nodejs`. But very likely it should be proposed in different repo in this org.

But, folks are _also_ free to continue to use this project as tested and maintained by volunteers, with all of the [resources](https://jupyterlab.readthedocs.io/en/stable/extension/extension_dev.html) for building and reusing extensions.
